src/typegate/src/runtimes/substantial.ts

```diff
@@ -0,0 +1,412 @@
+    typegate: Typegate
+  ) {
+    const basePath = path.join(typegate.config.base.tmp_dir, "artifacts");
+
+    for (const mat of materializers) {
+      if (
+        !["start", "stop", "send", "ressources", "results"].includes(mat.name)
+      ) {
+        continue;
+      }
+
+      const entryModulePath = await this.#getWorkflowEntryPointPath(
+        artifacts,
+        mat.data,
+        typegate
+      );
+
+      const { name: workflowName } = mat.data;
+      this.workflowFiles.set(workflowName as string, entryModulePath);
+      this.workflowResults.set(workflowName as string, []);
+      logger.info(`Resolved runtime artifacts at ${basePath}`);
+    }
+  }
+
+  async #getWorkflowEntryPointPath(
+    artifacts: Record<string, Artifact>,
+    matData: Record<string, unknown>,
+    typegate: Typegate
+  ) {
+    const entryPoint = artifacts[matData.file as string];
+    const deps = (matData.deps as string[]).map((dep) => artifacts[dep]);
+    const moduleMeta = {
+      typegraphName: this.typegraphName,
+      relativePath: entryPoint.path,
+      hash: entryPoint.hash,
+      sizeInBytes: entryPoint.size,
+    };
+    const depMetas = deps.map((dep) => {
+      return {
+        typegraphName: this.typegraphName,
+        relativePath: dep.path,
+        hash: dep.hash,
+        sizeInBytes: dep.size,
+      };
+    });
+
+    return await typegate.artifactStore.getLocalPath(moduleMeta, depMetas);
+  }
+
+  #eventResultHandlerFor(workflowName: string, runId: string) {
+    return async (result: Result<unknown>) => {
+      if (result.error) {
+        // All Worker/Runner non-user issue should fall here
+        // Note: Should never throw (typegate will panic), this will run in a worker
+        logger.error(
+          `result error for "${runId}": ${JSON.stringify(result.payload)}`
+        );
+        return;
+      }
+
+      const answer = result.payload as WorkerData;
+      logger.info(`"${runId}" answered: type ${JSON.stringify(answer.type)}`);
+
+      const startedAt = this.workerManager.getInitialTimeStartedAt(runId);
+
+      switch (answer.type) {
+        case "START": {
+          const ret = answer.data as WorkflowResult;
+
+          // Keep the backend updated with the latest emitted events
+          const run = ret.run;
+          run.operations = [...run.operations, ...this.#getStagedEvents(runId)];
+          const _run = nativeResult(
+            native.persistRun({
+              backend: this.backend,
+              run,
+            })
+          );
+
+          switch (Interrupt.getTypeOf(ret.exception)) {
+            case "SLEEP":
+            case "WAIT_ENSURE_VALUE":
+            case "WAIT_HANDLE_EVENT":
+            case "WAIT_RECEIVE_EVENT": {
+              await this.#workflowHandleInterrupts(workflowName, runId, ret);
+              break;
+            }
+            default: {
+              this.#workflowHandleGracefullCompletion(
+                startedAt,
+                workflowName,
```

Comment:
Note: This is analoguous to the try block on the poc and happens when a run is complete (with errors, with a result or with interrupts).